astropy/units/tests/test_quantity_non_ufuncs.py

```diff
@@ -2578,3 +2585,170 @@ def test_all_included(self):
     @needs_array_function
     def test_ignored_are_untested(self):
         assert IGNORED_FUNCTIONS | TBD_FUNCTIONS == untested_functions
+
+
+@pytest.mark.parametrize(
+    "functions",
+    sorted(
+        itertools.chain(FUNCTION_HELPERS.items(), DISPATCHED_FUNCTIONS.items()),
+        key=lambda items: items[0].__name__,
+    ),
+    ids=lambda items: items[0].__name__,
+)
+class TestFunctionHelpersSignatureCompatibility:
+    """
+    Check that a helper function's signature is *at least* as flexible
+    as the helped (target) function's. E.g., any argument that is allowed positionally,
+    or as keyword, by the target must be re-exposed *somehow* by the helper.
+    We explicitly allow helper's signature to be *more* flexible than the target signature
+    by allowing *args and **kwargs catch-all arguments, which we use to limit code
+    duplication, and also help with forward and backward compatibility.
+    See https://github.com/astropy/astropy/issues/15703
+    """
+
+    @staticmethod
+    def _have_catchall_argument(parameters, kind) -> bool:
+        return any(p.kind is kind for p in parameters.values())
+
+    @staticmethod
+    def _get_param_group(parameters, kinds: list) -> list[str]:
+        return [name for name, p in parameters.items() if p.kind in kinds]
+
+    def test_all_arguments_reexposed(self, functions):
+        target, helper = functions
+
+        try:
+            sig_target = inspect.signature(target)
+        except ValueError:
+            pytest.skip("Non Python function cannot be inspected at runtime")
+
+        params_target = sig_target.parameters
+        sig_helper = inspect.signature(helper)
+        params_helper = sig_helper.parameters
+
+        have_args_helper = self._have_catchall_argument(params_helper, VAR_POSITIONAL)
+        have_kwargs_helper = self._have_catchall_argument(params_helper, VAR_KEYWORD)
+
+        args_target = list(params_target.items())
+        args_helper = list(params_helper.items())
+
+        pos_target = 0
+        pos_helper = 0
+        while pos_target < len(args_target):
+            nt, pt = args_target[pos_target]
+            kt = pt.kind
+            if kt in (POSITIONAL_ONLY, POSITIONAL_OR_KEYWORD):
+                assert pos_helper < len(args_helper), (
+                    "helper's signature is too short; "
+                    "some arguments are not properly re-exposed"
+                )
+                nh, ph = args_helper[pos_helper]
+                if (kh := ph.kind) is not VAR_POSITIONAL:
```

Comment:
So for `ph.kind is VAR_POSITIONAL` one should fall through to the `if kt in (KEYWORD_ONLY, ...)` branch?

Reply:
I think this is incompatible with your suggestion to add a `continue`(which I think is correct !), isn't it ?

Reply:
Indeed, the `continue` would only be for this branch of `is not VAR_POSITIONAL`.

Reply:
Then yes, fall through is the intended behaviour here